File: go/plugins/onepassword/onepassword.go
// The onepassword plugin: 1Password, through a Connect server. Needs
// HTTPS. A port of typescript/plugins/onepassword.ts.
package onepassword

import (
	"net/http"
	"strconv"
	"strings"
	"sync"

	"github.com/voxgig/sekreto/go/plugins/httpjson"
	"github.com/voxgig/sekreto/go/sekreto"
)

// Provider reads 1Password, through a Connect server.
//
// The item titled api.token (titles keep their dots), in the named vault.
// The value is the field with purpose PASSWORD, or the field labelled
// `value`. A vault that cannot be found is an error - config names it, so
// its absence is a broken store, not a missing secret.
type Provider struct {
	Addr  string
	Token string
	Vault string
	// Guards the memoised state below: a Sekreto may resolve from several
	// goroutines, and a racy map read is either a crash or a zero value - a
	// MISS where the store does hold the secret, which falls through to a
	// weaker store.
	mu        sync.Mutex
	vaultid   string
	havevault bool
}

func (provider *Provider) auth() map[string]string {
	return map[string]string{"authorization": "Bearer " + provider.Token}
}

func (provider *Provider) resolvevault(addr string) (string, error) {
	want := provider.Vault
	if "" == want {
		return "", sekreto.Fail("sekreto: onepassword: no vault")
	}

	status, body, err := httpjson.Get(addr+"/v1/vaults", provider.auth())
	if nil != err {
		return "", err
	}

	list, is := body.([]any)
	if http.StatusOK != status || !is {
		return "", sekreto.Fail("sekreto: onepassword error: " + strconv.Itoa(status) + ": listing vaults")
	}

	for _, entry := range list {
		if want == httpjson.DigText(entry, "id") || want == httpjson.DigText(entry, "name") {
			return httpjson.DigText(entry, "id"), nil
		}
	}

	return "", sekreto.Fail("sekreto: onepassword: no vault named " + want)
}

func (provider *Provider) Lookup(name string) (string, bool, error) {
	if err := sekreto.CheckName(name); nil != err {
		return "", false, err
	}

	addr := strings.TrimSuffix(provider.Addr, "/")
	if "" == addr {
		return "", false, sekreto.Fail("sekreto: onepassword: no addr")
	}
	if err := sekreto.CheckAddr(addr); nil != err {
		return "", false, err
	}

	provider.mu.Lock()
	if !provider.havevault {
		vaultid, err := provider.resolvevault(addr)
		if nil != err {
			provider.mu.Unlock()
			return "", false, err
		}
		provider.vaultid = vaultid
		provider.havevault = true
	}
	vaultid := provider.vaultid
	provider.mu.Unlock()

	filter := httpjson.Escape(`title eq "` + name + `"`)
	status, body, err := httpjson.Get(
		addr+"/v1/vaults/"+vaultid+"/items?filter="+filter, provider.auth())
	if nil != err {
		return "", false, err
	}

	found, is := body.([]any)
	if http.StatusOK != status || !is {
		return "", false, sekreto.Fail("sekreto: onepassword error: " + strconv.Itoa(status) +
			": finding " + name)
	}

	if 0 == len(found) {
		return "", false, nil
	}

	itemid := httpjson.DigText(found[0], "id")
	status, body, err = httpjson.Get(
		addr+"/v1/vaults/"+vaultid+"/items/"+itemid, provider.auth())
	if nil != err {
		return "", false, err
	}

	if http.StatusOK != status {
		return "", false, sekreto.Fail("sekreto: onepassword error: " + strconv.Itoa(status) +
			": reading " + name)
	}

	fields, _ := httpjson.Dig(body, "fields").([]any)

	for _, field := range fields {
		if "PASSWORD" == httpjson.DigText(field, "purpose") {
			value := httpjson.Dig(field, "value")
			if nil == value {
				return "", false, nil
			}
			return httpjson.ToString(value), true, nil
		}
	}
	for _, field := range fields {
		if "value" == httpjson.DigText(field, "label") {
			value := httpjson.Dig(field, "value")
			if nil == value {
				return "", false, nil
			}
			return httpjson.ToString(value), true, nil
		}
	}

	return "", false, nil
}

func (provider *Provider) Describe() string {
	return "onepassword:" + provider.Vault
}

// Plugin is the `onepassword` provider kind, as a voxgig/plugin definition.
var Plugin = sekreto.ProviderPlugin("onepassword", func(spec *sekreto.ProviderSpec) (sekreto.Provider, error) {
	return &Provider{Addr: spec.Addr, Token: spec.Token, Vault: spec.Vault}, nil
})

File: java/plugins/com/voxgig/sekreto/plugins/Onepassword.java
// 1Password, through a Connect server, as a voxgig/plugin definition.
//
// A PLUGIN, NOT PART OF THE CORE: it opens a socket, and the core reads
// at most a local file. A chain that does not name this kind must not
// carry it, so the calling project imports this definition and hands it
// to Sekreto in the `plugins` option.
// See docs/design/plugin-providers.md.

package com.voxgig.sekreto.plugins;

import static com.voxgig.sekreto.Addr.checkaddr;
import static com.voxgig.sekreto.plugins.Httpjson.dig;
import static com.voxgig.sekreto.plugins.Httpjson.fetchjson;
import static com.voxgig.sekreto.plugins.Httpjson.headers;
import static com.voxgig.sekreto.plugins.Httpjson.trimslash;

import com.voxgig.sekreto.Provider;
import com.voxgig.sekreto.Sekreto;
import com.voxgig.sekreto.Sekreto.SekretoError;
import com.voxgig.sekreto.Support;
import com.voxgig.sekreto.plugins.Httpjson.Answer;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import voxgig.plugin.Definition;

/**
 * 1Password, through a Connect server.
 *
 * <p>The item titled `api.token` (titles keep their dots), in the named
 * vault. The value is the field with purpose PASSWORD, or the field
 * labelled `value`. A vault that cannot be found is an error - config
 * names it, so its absence is a broken store, not a missing secret.
 */
public final class Onepassword implements Provider {

  /** The `onepassword` kind: what `plugins` hands to Sekreto. */
  public static final Definition PLUGIN = Support.providerplugin("onepassword", spec ->
      new Onepassword(
          Support.text(spec.get("addr")), Support.text(spec.get("token")),
          Support.text(spec.get("vault"))));
  private final String addr;
  private final String token;
  private final String vault;

  private String vaultid;

  public Onepassword(String addr, String token, String vault) {
    this.addr = addr;
    this.token = token;
    this.vault = vault;
  }

  private Map<String, String> auth() {
    return headers("authorization", "Bearer " + (null == token ? "" : token));
  }

  private String resolvevault(String useaddr) {
    String want = null == vault ? "" : vault;
    if (want.isEmpty()) {
      throw new SekretoError("sekreto: onepassword: no vault");
    }

    Answer res = fetchjson("GET", useaddr + "/v1/vaults", auth(), null);

    if (200 != res.status || !(res.body instanceof List)) {
      throw new SekretoError("sekreto: onepassword error: " + res.status + ": listing vaults");
    }

    for (Object entry : (List<?>) res.body) {
      Object id = dig(entry, "id");
      if (want.equals(id) || want.equals(dig(entry, "name"))) {
        return String.valueOf(id);
      }
    }

    throw new SekretoError("sekreto: onepassword: no vault named " + want);
  }

  @Override
  public String lookup(String name) {
    Sekreto.checkname(name);

    String useaddr = trimslash(null == addr ? "" : addr);
    if (useaddr.isEmpty()) {
      throw new SekretoError("sekreto: onepassword: no addr");
    }
    checkaddr(useaddr);

    if (null == vaultid) {
      vaultid = resolvevault(useaddr);
    }

    String filter = Sigv4.uriescape("title eq \"" + name + "\"");
    Answer found = fetchjson(
        "GET", useaddr + "/v1/vaults/" + vaultid + "/items?filter=" + filter, auth(), null);

    if (200 != found.status || !(found.body instanceof List)) {
      throw new SekretoError(
          "sekreto: onepassword error: " + found.status + ": finding " + name);
    }

    List<?> items = (List<?>) found.body;
    if (items.isEmpty()) {
      return null;
    }

    Answer item = fetchjson(
        "GET", useaddr + "/v1/vaults/" + vaultid + "/items/" + dig(items.get(0), "id"),
        auth(), null);

    if (200 != item.status) {
      throw new SekretoError(
          "sekreto: onepassword error: " + item.status + ": reading " + name);
    }

    Object fields = dig(item.body, "fields");
    List<?> fieldlist = fields instanceof List ? (List<?>) fields : new ArrayList<>();

    for (Object field : fieldlist) {
      if ("PASSWORD".equals(dig(field, "purpose"))) {
        Object value = dig(field, "value");
        return null == value ? null : String.valueOf(value);
      }
    }
    for (Object field : fieldlist) {
      if ("value".equals(dig(field, "label"))) {
        Object value = dig(field, "value");
        return null == value ? null : String.valueOf(value);
      }
    }

    return null;
  }

  @Override
  public String describe() {
    return "onepassword:" + (null == vault ? "" : vault);
  }
}
